Additional collections features

damiani opened the pull request:
(Addresses some of the outstanding to-do items in https://github.com/tightenco/jigsaw/pull/59 and build on Adam's collection support...)

### Enhanced permalink support:
- `path` key is optional in `collections.php`; if not specified, collection item path will default to slugified filename
- Global `slugify()` helper added for easy use in `path` closure, helper functions, or Blade templates
- Shorthand for easily specifying common permalink structures
    - `path` key can be a string rather than a closure; if it's a string, any bracketed parameters like `{filename}` will be replaced with values
    - prepending a separator to a parameter name will slugify that parameter, using that separator (e.g. `{+filename}` will yield `the+slugified+filename`)
    - any dates from YAML front matter can be formatted using PHP date formatting codes, by following the parameter name with a pipe: `{date|Y/m/d}`, for example, yields `2016/08/29`.
    - if no parameters are included, the slugified filename will be appended by default. This allows users to simply specify a collection subdirectory: `'path' => 'posts'` will yield item URLs like `/posts/the-slugified-filename`.

Examples:
```
'path' => 'people'                                //   ./people/the-slugified-filename
'path' => 'people/{-filename}'                    //   ./people/the-slugified-filename
'path' => '{collection}/{date|Y}/{title}'         //   ./people/2016/The Item Title
'path' => '{collection}/{date|Y/m/d}/{+title}'    //   ./people/2016/08/29/the+item+title
```

### Default frontmatter variables:
- Add support for default frontmatter variables at the collection level, by specifying a `variables` key in `collections.php`:
```
'posts' => [
    'variables' => [
        'author' => 'Editorial Staff'
    ],
...
```

### Sorting collections:
- Add support for sorting collections by one or more criteria specified in `collections.php`.
- `sort` can contain a single variable name, or an array of multiple variables for a hierarchical sort.
- Sort order defaults to ascending; variable names can optionally be prepended with `+`.
- Variable names can be prepended with `-` for a descending sort order.
```
'posts' => [
    'sort' => '-date', 
...
```
or
```
'posts' => [
    'sort' => ['-date', '+author'], 
...
```

### Collection items contain a reference to next/previous items 
- Next/previous collection items are based on collection's default sort order
- e.g. `$jigsaw->item->next()->title` or `$jigsaw->item->previous()->path`

### Collections can exist without templates
- The `extends` directive can be omitted from the YAML front matter of a collection's items, allowing for collections that aren't tied to a particular template. 

### Access Jigsaw collections as Illuminate collections
- For example, `@foreach ($jigsaw->people->sortBy('age') as $person)` or `$jigsaw->products->sum('price')`
- Can even use helper functions defined in `config.php` or `collections.php`: 
```
@foreach ($jigsaw->products->sortByDesc(function ($data) { return $data->priceWithTax(); })
```
- Available anywhere on the site

### Additional pagination variables
- `$jigsaw->pagination->first`, `last`, `currentPage`, `totalPages`
- `$jigsaw->pagination->pages` can be iterated over to build numeric page links (`1` | `2` | `3`), or individual pages can be referenced as `$jigsaw->pagination->pages[1]`

### Helper functions
- Any helper functions defined in `collections.php` under the `helpers => []` key are available within Blade templates for a collection item; previously, they were only available in an index template when iterating over a collection. 
- Helper functions can also be defined globally in `config.php` and referenced in any Blade template as `$jigsaw->functionName($parameter)`

### Metadata
- available in Blade templates for all pages (including collection items and regular pages)
- `filename` displays current page's filename (without extension)
- `extension` displays current page's file extension (e.g. `md`, `blade.php`, etc.)
- `path` is path to current page, relative to site root
- `url` concatenates the site base url (if specified in `config.php`) with `path` for fully-qualified url to current page
- for collection items, `collection` displays the name of the collection an item belongs to

adamwathan commented:
Looks pretty good to me, the only thing I think is worth discussing is that slugifying seems to be applied even to custom permalinks. I think my gut feeling is to trust the person who writes the permalink function and not mess with their output, and if it causes problems let them deal with it, but maybe there's a good reason not to?

damiani commented:
Just end user simplicity, really. The same rationale for providing a default; I'd like to make collections as easy to use as possible out of the box, without having to do too much work in `collections.php`.

But it _is_ a bit magical, so perhaps it would be better just to document the availability of the `slugify` helper. This would have the added benefit of allowing users to specify a delimiter.

```
'permalink' => function ($data) {
    return slugify($data['title'], '+');
},
```

damiani commented:
In order to make collection and collection item handling a little easier and more flexible/extensible, I introduced a `Collection` parent object to manage a collection's details and to contain the child `CollectionItems`. This allows us to do things like store collection and item metadata only once and make them available everywhere, and to avoid double-parsing collection item markdown files.

For variables in Blade, I also switched the `$site` variable namespace to `$jigsaw`, and reference collection items within their single-item template as `$jigsaw->item->...`. One reason for doing this was that we can now reference helper functions defined in `collections.php` within a collection item template, but we needed a variable to hook those closures onto (so calling a helper function within a collection item's template looks like `$jigsaw->item->functionName()`). As a bonus, we can now also define helper functions globally in `config.php`, and reference those as well with `$jigsaw->functionName()`. Collections are referenced by their collection name, e.g. `$jigsaw->posts`, and can be iterated over anywhere.

A few other enhancements along the way include:
* enabling all variables to be referenced in views as either arrays or objects, and making these objects (and any nested arrays within them) fully iterable
* allowing collections to exist independent of a template (for lists of people, projects, etc.)
* allowing Jigsaw collections to be referenced as Illuminate collections, so we can do things like `@foreach ($jigsaw->people->sortBy('age') as $person)`
* added some extra pagination details (`first`, `last`, `totalPages`, `currentPage`, and numbered page links like `1 | 2 | 3`) 

Metadata support is in-progress; at the moment, you can reference a collection item's `filename` and `link` anywhere. I'm going to add support for non-collection pages as well, along with a `url` variable that will convert a `link` (which is relative) to a fully-qualified URL if the site's base url is defined in `config.php`.

adamwathan commented on src/Collection.php:
```diff
@@ -0,0 +1,90 @@
+        });
+        parent::__construct($this->addAdjacentItems($sortedItems));
+
+        return $this;
```
I can't figure out what the parent constructor call is for here, what's this meant to be doing?

damiani replied:
Since it's an Illuminate Collection, the call to the constructor is hydrating the collection with its children `CollectionItem`s (i.e. calling `$this->items = $this->getArrayableItems($items);`).

The `CollectionDataLoader` instantiates a new `Collection` using the named constructor `withSettings`, storing collection-specific settings, but not storing any items yet. The Loader then builds each `CollectionItem`—giving each of them a reference to its parent `Collection`—and then loads the items into the `Collection` with `loadItems` method.

An alternative would be to instantiate the `Collection` and load the items in the same call, which would then require looping over the items to store a reference to the parent `Collection` on each one.

adamwathan replied:
Could you just inline `$this->items = $this->getArrayableItems($items);` instead of the constructor call?

damiani replied:
Yep. I just called the constructor since, theoretically at least, Illuminate's Collection constructor could change at some point. Though obviously that's a remote concern.

adamwathan commented:
I feel like the code has gotten away from me a bit with this one, but based on the PR description the features all sound solid to me, so as long as this is working as expected I'm good with it 👍
